### ISSUE TITLE
Are you open to making TLS ciphers configurable?

### ISSUE THREAD

**Issue by yob** [heptio/contour#823]
Generally I'm reluctant to enter the custom cipher configuration game - the hunt for the perfect configuration to get an `A+` on qualys ssl labs is a mugs game.

However, we've hit a situation where we'd prefer to sacrifice some backwards compatibility to get perfect forward secrecy in place for all connections. Rats.

Are you open to a patch to make the ciphers enabled in envoy configurable?

The following comment in the ingress route design doc (`design/ingressroute-design.md`) sounds promising:

> other properties like cipher suites may be added later

We're currently still using contour 0.7 with traditional Ingress objects, but I guess we'll have to switch to using IngressRoute eventually, and if they make configurable ciphers an option then that'd be a bonus.

**Comment by davecheney** [comment]
Yes, I think we can do that in both ingress via an annotation and ingress route in the virtual host.tls spec.

**Comment by yob** [comment]
That's good news.

The envoy cipher config is the standard list of cipher labels as a string. Is your preference to expose that directly so contour users can can provide a cipher string?

**Comment by davecheney** [comment]
My first preference is to fix whatever ssl labs is complaining about without an option to change it. 

I’m not sure that is achievable, so the fallback is to choose a reasonable default then add a list argument to the tls spec where an alternative list can be supplied.

**Comment by yob** [comment]
I had a bit of a poke at this. Customising the list of ciphers passed to envoy is simple enough, but making the feature robust and not a giant foot cannon was difficult.

It seems that passing an unrecognised cipher to envoy results in envoy not starting a TLS listener on the the requested port, and the supported list of ciphers depends on the vendored cersion of boringssl in envoy. It was *very* easy for me to end up with nothing listening on (8)443 without obvious feedback about why.

Does anyone know if it's possible to query envoy for a list of supported ciphers? if so, we could compare that list against the user provided list and return a nice error if the user has requested one that boringssl doesn't support.

**Comment by davecheney** [comment]
At the moment Envoy and Contour are tightly coupled so I would like to avoid having to sense the ciphers supported by envoy, present them to the user, have them adjust them on a per ingress basis, presumably with some filtering from the administrator. Apart from the fiddlyness of this process, I don't think its actually solving the problem.

The problem users and administrators want is for people to not use insecure ciphers. I would prefer to see the list of ciphers supported by contour kept up to date so that all users can benefit. We did this before when we turned off TLS/1.1 support, nobody seemed to actively want to use insecure ciphers and the argument that they need to support old busted pieces of equipment get weaker over time.

In summary, even though we _could_ present the option for users and administrators to control their cipher set, I would prefer that we did not, instead Contour should configure Envoy so that nobody is using insecure ciphers.

**Comment by yob** [comment]
I'm on board with being opinionated, although I'm worried that my opinion about what ciphers are secure may be ill-informed. Particularly over the long term where the recommended ciphers will evolve and *someone* (possibly not me) in the contour community will have to take on the responsibility of keeping contours cipher list up to date.

I guess I could prepare a PR that selects the ciphers required to get an A on qualys, and we could have a discussion about the compatibility impacts. We'd have to be prepared for a few folks to find upgrading contour breaks compatibility with a device they care about though.

**Comment by davecheney** [comment]
Qualys or ssl labs would be a good start. I'm not an expert on this either,
but I know that ssllabs A+ is a goal for some teams, so shooting for that
(the current cipher list isn't that short of A+ at the moment) sounds like
a reasonable goal

**Comment by davecheney** [comment]
Based on my two most recent comments I’m marking this as help wanted for 0.12. 

If you want to work on it the spec is to revise the list of permitted ciphers to raise the ssl labs score. 

No design doc is needed but I expect a detailed commit message detailing the decisions and trade offs of applying the PR. 

Thank you.

**Comment by yob** [comment]
Thanks @davecheney - I've had a crack at something in #1011
